db/judge/SolidityJudge.js

```diff
@@ -46,46 +45,74 @@ class SolidityJudge extends BaseJudge {
-    const contract = new this.web3.eth.Contract(ABI)
-
-    if (!this.contractHas(contract.methods, props.methods)) {
-      throw new FunctionalityError(`Contract methods are missing!. Expecting: ${JSON.stringify(props.methods)}`)
-    }
-
-    if (!this.contractHas(contract.events, props.events)) {
-      throw new FunctionalityError(`Contract events are missing!. Expecting: ${JSON.stringify(props.events)}`)
-    }
-
-    const accounts = await this.web3.eth.getAccounts()
-
-    const instance = await contract
-      .deploy({ data: bytecode, arguments: props.args })
-      .send({ from: accounts[0], gas: 1000000 })
-
-    this.contractInstance = instance
-
-    return instance
+  deploy (name, source, props) {
+    return new Promise(async (resolve, reject) => {
+      try {
+        await this.web3.eth.net.isListening()
+      } catch (e) {
+        reject(new BlockchainConnectionError())
+      }
+
+      const input = {
+        language: 'Solidity',
+        sources: {
+          [`${name}.sol`]: {
+            content: source
+          }
+        },
+        settings: {
+          outputSelection: {
+            '*': {
+              '*': [ '*' ]
+            }
+          }
+        }
+      }
+
+      try {
+        let compiled = solc.compile(JSON.stringify(input))
+        compiled = JSON.parse(compiled)
+
+        if (compiled.errors && compiled.errors.length > 0) {
+          reject(new CompilationError(compiled.errors))
+        }
+
+        if (Object.keys(compiled.contracts[`${name}.sol`])[0] !== name) {
+          reject(new CompilationError([{ formattedMessage: `Contract name should be ${name}` }]))
+        }
+
+        const compiledContract = compiled.contracts[`${name}.sol`][name]
+        const contractMethods = compiledContract.abi.filter(method => method.type === 'function').map(method => method.name)
+        const contractEvents = compiledContract.abi.filter(method => method.type === 'event').map(method => method.name)
+
+        if (!this.contractHas(contractMethods, props.methods)) {
+          reject(new FunctionalityError(`Contract methods are missing!. Expecting: ${JSON.stringify(props.methods)}`))
+        }
+
+        if (!this.contractHas(contractEvents, props.events)) {
+          reject(new FunctionalityError(`Contract events are missing!. Expecting: ${JSON.stringify(props.events)}`))
+        }
+
+        const ABI = compiledContract.abi
+        const bytecode = compiledContract.evm.bytecode.object
+        const contract = new this.web3.eth.Contract(ABI)
+
+        const accounts = await this.web3.eth.getAccounts()
+
+        contract
+          .deploy({ data: bytecode, arguments: props.args })
+          .send({ from: accounts[0], gas: 2000000 })
+          .on('receipt', (receipt) => {
+            const instance = new this.web3.eth.Contract(ABI, receipt.contractAddress)
+            this.contractInstance = instance
+            resolve(instance)
+          })
+          .on('error', reject)
```

Comment:
It looks to me as if you've switched from async/await to resolve/reject due to requirement to capture the error at this point. I think you should separate the real contract deployment function into a separate function (say, `_deploy`) that follows the resolve/reject paradigm and keep this function (`deploy`) clean with async/await. Then `_deploy` can call `reject` on error and `resolve` on receipt, and `deploy` can `await _deploy(...)`